Relocatable OCaml - explicit-relative paths in `ld.conf`

dra27 opened the pull request:
This is the first of three PRs which implement Relocatable OCaml as proposed in ocaml/RFCs#53. This PR specifically deals with changes to the handling of `ld.conf`.

The key changes are:
- A new `configure` option `--with-stublibs` added to address ocaml/opam-repository#16406 which specifies a custom line to be added to the top of `ld.conf`. This new flag is intended for use in `opam` and provides a mechanism to eliminate the setting of `CAML_LD_LIBRARY_PATH` by the compiler packages in opam-repository.
- Lines in `ld.conf` beginning with an explicit reference to the current or parent directory (i.e. `./` or `../`) are interpreted relative to the directory containing `ld.conf`.
- Previously, if `OCAMLLIB` (or `CAMLLIB`) were set in the environment, then the `ld.conf` was loaded from the directory referred to by this environment variable, and the `ld.conf` file located in the configured Standard Library location was ignored. The runtime now loads `ld.conf` from all of these locations, processing them in the same order.

I have presented all of these changes as one PR because while the individual parts of it can be separated, they all alter the same part of the runtime and don't recombine cleanly. While overhauling this infrequently-altered section of the runtime, I propose some further alterations:
- The processing of `ld.conf` is made line-ending agnostic - Unix and Cygwin builds now consistently ignore CR characters at the end of each line[^1]
- Static builds of the compiler no longer create a `stublibs` directory, nor add it to `ld.conf`
- The C implementation of the processing for `ld.conf` is exposed as a primitive which is now used by `ocamlc`, instead of having both C and OCaml implementations of the "same" algorithm. My reason for proposing this is twofold: firstly, the two implementations even before this PR are inconsistent (as the special cases in the testsuite reveal) and, secondly, with the other changes proposed here, the two implementations become considerably larger, and thus the benefit of sharing the code is bigger. The commit unifying the two implementations is intentionally towards the end of the series (cf. the deletions in `bytecomp/dll.ml` in the penultimate commit, versus the deletions in `bytecomp/dll.ml` for the overall PR)

As far as I know, the only thing apart from OCaml which parses `ld.conf` is `ocamlfind`, and the patch for that isn't particularly involved (see, for example, [dra27/ocamlfind#7bd58cb](https://github.com/dra27/ocamlfind/commit/7bd58cb14a84a82223091098af32a4da471b3984)). The compiler drivers use the prefix `+` to refer to the Standard Library (e.g. in `-I +unix`). Co-opting this notation for `ld.conf` doesn't work in combination with the change to load `ld.conf` from all possible locations, since `+` refers to the _effective_ Standard Library location. Another possibility is to refer to the _prefix_ by a given name, for example, having the default being `$/lib/ocaml/stublibs` and `$/lib/ocaml`. Establishing the idea of `$` may be useful for other tools; rooting the names of files in the _prefix_ rather than the directory containing the file.

[^1]: The nefarious user who absolutely must have directories in `ld.conf` which end with CR characters is able to encode this by including a trailing slash after the name...

dra27 committed:
Load ld.conf from all possible places

Before, the first ld.conf found from $OCAMLLIB, $CAMLLIB or the
preconfigured standard library location was loaded. Now all of these
are loaded.

dra27 committed:
Harden the parsing of ld.conf w.r.t. load and CRLF

dra27 committed:
ld.conf-relative path interpretation

Explicit relative paths in ld.conf are now interpreted relative to the
directory containing ld.conf.

dra27 committed:
Generate ld.conf using relative paths

dra27 committed:
Remove caml_get_stdlib_location

The function was only ever added to share the logic between dynlink.c
and startup_byt.c - now that dynlink.c doesn't require it, move the
function to startup_byt.c and make it internal again.

dra27 committed:
Add --with-stublibs configure option

--with-stublibs is intended for use in opam and allows an additional
directory to be added at the top of $(ocamlc -where)/ld.conf.

dra27 committed:
Use caml_parse_ld_conf in ocamlc

Eliminate the need for two implementations of the parsing logic for
ld.conf by sharing the C implementation (which must exist, since it's
part of bytecode startup) with the bytecode compiler, replacing
Dll.ld_conf_contents

dra27 committed:
Don't add a double-separator when locating ld.conf

When calculating the full path for ld.conf, the runtime unconditionally
concatenated "/ld.conf". This is harmless when the separators appear in
the middle of a path ("/usr/local/lib/ocaml//ld.conf" is equivalent to
the version with only single slashes), but it is technically incorrect
for two corners cases with OCAMLLIB and CAMLLIB:

- if either is explicitly set to "/" then "//ld.conf" is _not_ the same
  file as "/ld.conf". This is mildly relevant on Windows and Cygwin
  where the two initial slashes (including as "\/" for native Windows)
  will be interpreted as a UNC path
- if either is explicitly blank, then "ld.conf" (i.e. ld.conf in the
  current directory) is a less illogical file to open than "/ld.conf"

dra27 committed:
Tidy installation of static builds

- $libdir/stublibs is no longer created for a --disable-shared build
- When $libdir/stublibs is not created, it is also not added to ld.conf

damiendoligez commented on runtime/dynlink.c:
```diff
@@ -78,65 +78,155 @@ static c_primitive lookup_primitive(const char * name)
+}
+
+/* Exposes caml_parse_ld_conf as a primitive for the bytecode compiler, saving
+   the duplication of the logic with the bytecode compiler. */
```
```suggestion
   the duplication of the logic within the bytecode compiler. */
```

damiendoligez commented on runtime/dynlink.c:
```diff
@@ -78,65 +78,155 @@ static c_primitive lookup_primitive(const char * name)
+      libroot = caml_stat_strdup_os(locations[i]);
+      size_t libroot_length = strlen_os(libroot);
+      if (libroot_length > 1 && Is_separator(libroot[libroot_length - 1]))
+        libroot[libroot_length - 1] = '\0';
```
If you want to handle the corner cases of :
- several slashes at the end of a path
- the user putting OCAMLLIB at the file system root 😂
- both of the above

```suggestion
      while (libroot_length > 0 && Is_separator(libroot[libroot_length - 1]))
        libroot[--libroot_length] = '\0';
```

dra27 replied:
Technically-speaking (and more by luck than judgement) the previous code "correctly" preserved `//` as a root, where this version now treats it as `/`. A user with `OCAMLLIB=/` is already in a lot of trouble; the Cygwin user who ended up with `OCAMLLIB=//` and a host on the network named `ld.conf` was in even more trouble, and may come back and thank us for changing this...

jonahbeckford commented on runtime/caml/dynlink.h:
```diff
@@ -41,11 +41,10 @@ extern void caml_build_primitive_table_builtin(void);
-extern char_os * caml_parse_ld_conf(void);
+/* If found, parse $OCAMLLIB/ld.conf, $CAMLLIB/ld.conf and stdlib/ld.conf in
+   that order and add the lines read to table. */
+extern char_os * caml_parse_ld_conf(const char_os * stdlib,
```
FYI: I did a GitHub search for `caml_parse_ld_conf` since adding arguments to the argument list of `caml_parse_ld_conf` is veering into C undefined behavior (if client code accesses an old `caml_parse_ld_conf` in shared library, or worse if vice versa). No one outside of startup_byt.c is using it, so it looks safe to change.

dra27 replied:
Indeed, it's guarded by `CAML_INTERNALS` as well (it's actually only been exposed at all since #9284 in 4.13.0 in order to implement `ocamlrun -config` ... which was very very early work on Relocatable!)

jonahbeckford commented on runtime/startup_byt.c:
```diff
@@ -376,6 +376,15 @@ static int parse_command_line(char_os **argv)
+{
+  const char_os * stdlib;
+  stdlib = caml_secure_getenv(T("OCAMLLIB"));
+  if (stdlib == NULL) stdlib = caml_secure_getenv(T("CAMLLIB"));
```
For portable consistency, can we check for NULLs and empty strings? (Unless it is documented somewhere that empty OCAMLLIB stops search for CAMLLIB and OCAML_STDLIB_DIR, which would be weird).

dra27 replied:
I've tackled this in #14247 (the specific bit can be previewed in dra27/ocaml#227, although this function moves about a bit, so it's in `dynlink.c` in that PR still!). My rationale was to preserve the "status quo" for in this PR, just because it's not strictly necessary to fix it but, as you can see from the follow-up (and from dra27/ocaml#225), I firmly agree that we should treat NULL and empty as equivalent!

jonahbeckford commented:
LGTM, with minor comment. I didn't review the test suite since I'm unfamiliar with that.

dra27 commented:
Thanks, @jonahbeckford!

> I didn't review the test suite since I'm unfamiliar with that

Some more detail for these changes. They're slightly better looked at commit-by-commit:
- 2e8737095b42257b8cf8114f60299fd41704d219 (ld.conf-relative path interpretation): the change in `test_ld_conf.ml` reflects the new interpretations of some of the lines (e.g. `.` now being interpreted as libdir)
- a2e900a61dc75a1ba72396ea2a6492d141e99fe6 (Generate ld.conf using relative paths): this affects `testRelocation.ml` because `ld.conf` no longer contains absolute paths. Similarly, `testLinkModes.ml` is affected because after renaming the directory containing the installed compiler, `ld.conf` is now correctly processed by the runtime (since the `.` and `./stublibs` entries in it are both still correct)
- 9062f609846a09380a783991cbc36695ca166960 (Load ld.conf from all possible places): this commit alters the runtime to load up to three `ld.conf` files, instead of one (i.e. `$OCAMLLIB/ld.conf`, `$CAMLLIB/ld.conf` and `OCAML_STDLIB_DIR/ld.conf` are now all loaded, if they exist). That necessarily affects the tests for shadowing in `test_ld_conf.ml`.
- 5b339f888c0cb1af21bf5bfbd3e9ec938736958f (Harden the parsing of ld.conf w.r.t. load and CRLF): makes the processing of `ld.conf` consistent w.r.t. CRLF, which consequently affects the tests in `test_ld_conf.ml` which showed the previous non-portable behaviour
- 1ddfabcacac2596abcfc2ca3ef0a8e0568835a88 (Use caml_parse_ld_conf in ocamlc): the implementations of the parsing in the `dll.ml` and `dynlink.c` differed in various corner cases, all of which were exercised in `test_ld_conf.ml` - the change therefore is the removal of all the special cases 🥳

dra27 commented:
Surfacing a brief discussion on caml-devel about a design-decision affecting this. Non-blank lines in `ld.conf` are each one of three kinds of path:
- Absolute (`/usr/…`, `C:\Users\…`, `\\DC0\Home\…`)
- Explicit-relative (`.`, `..`, `./…` or `../…`)
- Implicit (`foo`, `bar/…`, etc.)

In OCaml today, only absolute directories have a sensible semantics, so the core idea in this PR is to add a sensible semantics to either or both of the other two lines (explicit-relative, implicit), on the basis that they're presently not in use.

A _complete_ test for an absolute path is quite difficult to get right on Windows, which is necessarily needed to determine if a line is an implicit path. However, the explicit-relative test is straightforward, which is the main reason I opted just to interpret explicit-relative directories with this new mechanism (it is also the case that opam-bin uses a `+dir` notation in its patches for `ld.conf`, and it meant that this would not interfere with that - although that's a relatively minor concern, as I'm sure opam-bin will stop patching the compiler when there's upstream support for relocation anyway).

In this PR right now, we have explicit-relative paths interpreted relative to where `ld.conf` is found. That means, in opam, the content of the file will be:
```
../stublibs
./stublibs
.
```

There are two changes either or both of which could be made. Firstly, the paths could be instead be interpreted relative to `--prefix` (that requires a little bit of rework to pull in `--with-relative-libdir` from #14244 or something like it). The file would then be:
```
./lib/stublibs
./lib/ocaml/stublibs
./lib/ocaml
```
The second change would be to recognise both explicit-relative and implicit in the same way, allowing
```
lib/stublibs
lib/ocaml/stublibs
lib/ocaml
```
The decision in this PR has a small bearing on `ocamlfind` (but as far as I can tell for nothing else in the ecosystem), as `findlib.conf` needs to record `destdir` and `path`. In my draft for that at https://github.com/dra27/ocamlfind/pull/2, `.` is being interpreted relative to `findlib.conf`, giving:
```
destdir="."
path="./ocaml:."
```
but which could therefore be:
```
destdir="./lib" # Or just destdir="lib"
path="./lib/ocaml:./lib" # Or just path="lib/ocaml:lib"
```

hhugo commented on bytecomp/dll.ml:
```diff
@@ -141,12 +141,36 @@ let synchronize_primitive num symb =
+          if line = "" then
+            ""
+          else
+            let len = String.length line in
```
I've tried to make the logic easier to understand. I hope I've managed to preserve all aspect. 
```ocaml
if Filename.is_relative line && not (Filename.is_implicit line)
then
  match line with
  | "." -> Config.standard_library
  | _ -> Filename.concat Config.standard_library line
else line
```

dra27 replied:
(note that this function gets completely deleted in 1ddfabcacac2596abcfc2ca3ef0a8e0568835a88, where `Dll` co-opts the C implementation).

It's different for `line = "./stublibs"`, I think - the previous version (and the version in C) produces `<libdir>/stublibs`, but the newer version produces `<libdir>/./stublibs`. It's pedantic, but I was trying where possible to avoid the stray `/./` bits being introduced, given that the default entries are `.` and `./stublibs`!

jonahbeckford commented:
> Some more detail for these changes. They're slightly better looked at commit-by-commit:
> ...

Thanks, that was helpful for me to review the testsuite. It all looks good to me.

dra27 commented:
Rebased - and with one additional change inspired by https://github.com/ocaml/ocaml/pull/14244#discussion_r2393971938. I've renamed `--with-stublibs` to the slightly clearer (I think) `--with-additional-stublibsdir` (stublibsdir vs libdir, etc.) and then made `../stublibs` the default for that, which also makes the opam file neater. As in #14244, the changes are in extra "Review" commits

dra27 commented:
Another follow-up from https://github.com/ocaml/ocaml/pull/14246#issuecomment-3435923894. The last commit in this series corrected the over-installation of the native stub libraries added in #11828. That's only a partial fix (they shouldn't be built either) and it doesn't belong in this PR.

damiendoligez commented on Changes:
```diff
@@ -66,6 +66,17 @@ Working version
+  unaltered. Additionally, ld.conf is loaded from all of $OCAMLLIB/ld.conf,
+  $CAMLLIB/ld.conf and standard_library_default/ld.conf rather than just the
+  first one found. ld.conf files with CRLF line endings are now consistently on
+  both Windows and Unix.
```
Aren't we missing a verb here?

dra27 commented:
There's a few points which I'll go through later today and then squash the Review commits in prior to merging

dra27 commented:
Review comments addressed (cf. the Compare button above) - squashing the review commits back in.

dra27 commented:
Thank you to everyone for the reviewing - this is having a final run through [precheck#1080](https://ci.inria.fr/ocaml/job/precheck/1080/) and, assuming no unexpected CI results is ready to go!
